clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java

```diff
@@ -185,24 +186,35 @@ public synchronized void unsubscribe() {
         final Map<TopicPartition, List<ConsumerRecord<K, V>>> results = new HashMap<>();
 
         for (Map.Entry<TopicPartition, List<ConsumerRecord<K, V>>> entry : this.records.entrySet()) {
-            if (!subscriptions.isPaused(entry.getKey())) {
+            if (!subscriptions.isPaused(entry.getKey()) && assignment().contains(entry.getKey())) {
                 final List<ConsumerRecord<K, V>> recs = entry.getValue();
                 for (final ConsumerRecord<K, V> rec : recs) {
                     if (beginningOffsets.get(entry.getKey()) != null && beginningOffsets.get(entry.getKey()) > subscriptions.position(entry.getKey())) {
-                        throw new OffsetOutOfRangeException(Collections.singletonMap(entry.getKey(), subscriptions.position(entry.getKey())));
+                        if (subscriptions.hasDefaultOffsetResetPolicy()) {
+                            subscriptions.requestOffsetReset(entry.getKey());
+                            resetOffsetPosition(entry.getKey());
+                        } else {
+                            throw new OffsetOutOfRangeException(Collections.singletonMap(entry.getKey(), subscriptions.position(entry.getKey())));
+                        }
                     }
+                }
+                final List<ConsumerRecord<K, V>> recsAfterPosition = recs.stream()
+                        .filter(rec -> rec.offset() >= subscriptions.position(entry.getKey()))
+                        .collect(Collectors.toList());
 
-                    if (assignment().contains(entry.getKey()) && rec.offset() >= subscriptions.position(entry.getKey())) {
-                        results.computeIfAbsent(entry.getKey(), partition -> new ArrayList<>()).add(rec);
-                        subscriptions.position(entry.getKey(), rec.offset() + 1);
-                    }
+                for (final ConsumerRecord<K, V> rec : recsAfterPosition) {
+                    results.computeIfAbsent(entry.getKey(), partition -> new ArrayList<>()).add(rec);
+                    subscriptions.position(entry.getKey(), rec.offset() + 1);
                 }
             }
         }
-        this.records.clear();
         return new ConsumerRecords<>(results);
     }
 
+    public synchronized void clearRecords() {
+        this.records.clear();
```

Comment:
Previously the records were consumed after every poll. Now I think the intent is to treat the records collection as representing the backing log in Kafka. Is that about right? 

Assuming so, I wonder if we can make the representation a little clearer. We currently have separate collections for `beginningOffsets`, `endOffsets`, and `records`. Perhaps we can consolidate all of them. For example, in pseudocode, we could have something like this:

```java
class MockLogData {
  List<ConsumerRecord> log;

  long startOffset() { return log.first.offset(); }
  long endOffset() { return log.last.offset() + 1; }
  List<ConsumerRecord> fetch(long offset) throws OffsetOutOfRangeException; 
}
```

Then we could replace the three collections with a single `Map<TopicPartition, MockLogData>`.

Reply:
> Now I think the intent is to treat the records collection as representing the backing log in Kafka. Is that about right?

Yes.

I think something like MockLogData makes sense. We can maybe get rid of `updateBeginningOffsets` and replace it with the ability to delete single offsets from `records`, but we might need to keep `addEndOffset` around. It seems to be intended for testing the case where the code being tested calls `endOffsets` more than once, and the test wants to return something different on each call to simulate records being added concurrently.

I'll try to make something like MockLogData and will let you know when it's ready for another look.